src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -1229,23 +1233,126 @@ private int SkipBlockComment(int i)
+                case 'r': return "\r";
+                case 't': return "\t";
+                case 'u': return ScanUnicodeEscapeSequence();
+                case 'v': return "\v";
+                default: return c.ToString();
+            }
+        }
+
+        private string ScanUnicodeEscapeSequence()
+        {
+            int escSeqStartIndex = _currentIndex - 2;
+            int maxNumberOfHexDigits = 6;
+            bool isTerminated = false;
+
+            char c = GetChar();
+            if (c != '{')
+            {
+                UngetChar();
+
+                IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                return s_unknownUnicodeChar;
+            }
+
+            // Scan hex chars after the Unicode escape sequence start.
+            var sb = GetStringBuilder();
+            int i;
+            for (i = 0; i < maxNumberOfHexDigits; i++)
+            {
+                c = GetChar();
+
+                // Sequence has been terminated.
+                if (c == '}')
+                {
+                    if (i == 0)
+                    {
+                        // Sequence must have at least one hex char.
+                        IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                        ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                        return s_unknownUnicodeChar;
+                    }
+
+                    isTerminated = true;
+                    break;
+                }
+
+                if (!c.IsHexDigit())
+                {
+                    UngetChar();
+
+                    ReportError(_currentIndex, () => ParserStrings.MissingUnicodeEscapeSequenceTerminator);
+                    return s_unknownUnicodeChar;
+                }
+
+                sb.Append(c);
+            }
+
+            if (!isTerminated)
+            {
+                c = GetChar();
+                if (c != '}')
+                {
+                    UngetChar();
+                    ReportError(_currentIndex,
+                        i == maxNumberOfHexDigits && c.IsHexDigit()
+                            ? (Expression<Func<string>>)(() => ParserStrings.TooManyDigitsInUnicodeEscapeSequence)
+                            : () => ParserStrings.MissingUnicodeEscapeSequenceTerminator);
+                    return s_unknownUnicodeChar;
+                }
+            }
+
+            string hexStr = GetStringAndRelease(sb);
+
+            int unicodeValue = int.Parse(hexStr, NumberStyles.AllowHexSpecifier, NumberFormatInfo.InvariantInfo);
+            if (unicodeValue <= Char.MaxValue)
+            {
+                return ((char)unicodeValue).ToString();
+            }
+            else if (unicodeValue <= 0x10FFFF)
+            {
+                return Char.ConvertFromUtf32(unicodeValue);
```

Comment:
>  using surrogate characters is rejected as rudiment

I disagree on this one.  There's no good reason to disallow the specification of **any** Unicode character.  Both C# and JavaScript allow this.

> Accept 1 to 8 digits
Allowing up to 8 digits seems enticing from a future proofing perspective but it doesn't really help in that regard.  Any value over 0x10FFFF, which only requires six digits, is invalid and will generate a parse error.   **If** Unicode is updated to support characters above 0x10FFFF the parser would have to be modified to adjust the 0x10FFFF limit check - as would the [C# parser](https://github.com/dotnet/roslyn/blob/94ca919c6b4c94857dc5d58df4e6e10cb8d55c3c/src/Compilers/CSharp/Portable/Parser/SlidingTextWindow.cs#L440) BTW.  At the time the parser is modified to bump that limit (if it ever happens) then we can bump the maxNumberHexDigits limit as well.  This is actually an advantage of using the variable width w/braces form.  Increasing the maxNumberHexDigits is not a breaking change as it would be for the fixed width form.